### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by AniVerma17**
- Add missing upload API parameters in the documentation
- Add API documentation for the new features in development

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -328,66 +349,165 @@ The complete list of transformations supported and their usage in ImageKit can b
...
+| overlayImageBackground        | overlayImageBackground(overlayImageBackground: String)                                                                                                                                                                                                                                                                                               | oibg                    |
+| overlayImageBorder            | overlayImageBorder(borderWidth: Int, borderColor: String)                                                                                                                                                                                                                                                                                            | oib                     |
```
Comment (imagekitio) [inline README.md]: Please remove all `overlay` related parameter. In future, all these parameters will be removed and new powerful layer syntax will be the only way to use overlays. In SDKs, we will expose `raw` parameter and allow users to set any string.

File: README.md [hunk 2]
```diff
@@ -42,34 +42,44 @@ implementation 'com.github.imagekit-developer:imagekit-android:<VERSION>'
```
Comment (imagekitio) [inline README.md:67]: Add a table of index and link all important sections of this readme at the top like this - https://github.com/imagekit-developer/imagekit-nodejs#table-of-contents

File: README.md [hunk 3]
```diff
@@ -328,66 +349,165 @@ The complete list of transformations supported and their usage in ImageKit can b
...
+| original                      | original(flag: Boolean)                                                                                                                                                                                                                                                                                                                              | orig                    |
+| Raw param string              | raw(params: String)                                                                                                                                                                                                                                                                                                                                  | -                       |
 
 </details>
 
-### File Upload
+### Responsive image loading
+To automatically set the dimensions and pixel ratio of the image , call `ImageKit.getInstance().url(...).setResponsive(...)` with a set of parameters defined below:
```
Comment (imagekitio) [inline README.md:389]: Still missing key information regarding how responsive images will actually work and what can user expect.

File: README.md [hunk 4]
```diff
@@ -328,66 +349,165 @@ The complete list of transformations supported and their usage in ImageKit can b
...
+    .setAdaptiveStreaming(StreamingFormat.HLS,
+        Arrays.asList(240, 360, 480, 720, 1080, 1440, 2160)
+    )
+    .create();
+```
+
+## File Upload
 The SDK provides a simple interface using the `ImageKit.getInstance().uploader().upload(...)` method to upload files to the ImageKit Media Library. It accepts all the parameters supported by the [ImageKit Upload API](https://docs.imagekit.io/api-reference/upload-file-api/client-side-file-upload#request-structure-multipart-form-data).
 
 Make sure that you have specified `authenticationEndpoint` during SDK initialization. The SDK makes an HTTP GET request to this endpoint and expects a JSON response with three fields i.e. `signature`, `token`, and `expire`.
```
Comment (imagekitio) [inline README.md]: Let's change this. In this new SDK:
* Use [secure client side upload - V2](https://docs.imagekit.io/api-reference/upload-file-api/secure-client-side-file-upload)
* Let user of the SDK generate token however they wish and make `token` parameter compulsory in upload API payload. Don't ask for `authenticationEndpoint` and internally hit that endpoint to get `tokent` like we do right now. I can explain more about this on the call and why we are changing this.

Reasons for changing this:
* V2 uses JWT based authentication which verifies the whole payload content.
* Our SDK call `authenticationEndpoint` without any custom header, so for users its still unsecured endpoint at their end. They want to handle authentication themself, so SDK should not get in the way.

File: README.md [hunk 5]
```diff
@@ -396,40 +516,210 @@ Make sure that you have specified `authenticationEndpoint` during SDK initializa
...
-  imageKitCallback = object: ImageKitCallback {
+    file = image,
+    fileName = "sample-image.jpg",
+    useUniqueFilename = true,
+    tags = ["demo"],
+    folder = "/",
+    isPrivateFile =  false,
+    customCoordinates = "",
+    responseFields = "",
+    extensions = listOf(mapOf("name" to "google-auto-tagging", "minConfidence" to 80, "maxTags" to 10)),
```
Comment (imagekitio) [inline README.md:527]: Is there not a better way to take this information? By better I mean more strongly typed like we do all things in Java/Kotlin.

File: README.md [hunk 6]
```diff
@@ -396,40 +516,210 @@ Make sure that you have specified `authenticationEndpoint` during SDK initializa
...
+| requiresDeviceIdle(requiresIdle: Boolean)          | UploadPolicy.Builder | Sets whether the device needs to be idle for the upload request. Defaults to `false`.                                                                                                                                            |
+| setMaxRetries(count: Int)                          | UploadPolicy.Builder | Sets the maximum number of retries for the upload request. Negative value will throw an `IllegalArgumentException`. Defaults to 5.                                                                                               |
+| setRetryBackoff(interval, BackoffPolicy policy)    | UploadPolicy.Builder | Sets the backoff interval in milliseconds and policy (`UploadPolicy.BackoffPolicy.LINEAR` or `UploadPolicy.BackoffPolicy.EXPONENTIAL`) for retry attempts. Defaults to interval of 10000ms and policy of `BackoffPolicy.LINEAR`. |
+
+Example code
```
Comment (imagekitio) [inline README.md:566]: In addition to code, explain what this does in a paragraph.

File: README.md [hunk 7]
```diff
@@ -396,40 +516,210 @@ Make sure that you have specified `authenticationEndpoint` during SDK initializa
...
+```kotlin
+val preprocessor = ImagePreprocessor.Builder()
+    .limit(1280, 720)
+    .format(Bitmap.CompressFormat.WEBP)
+    .rotate(45f)
+    .build()
+```
+### Video preprocessing
+The `VideoPreprocessor` class encapsulates a set of methods to apply certain transformations to an image before uploading.
```
Comment (imagekitio) [inline README.md]: Incorrect copy paste and mention of `image` in video section. Also please describe this with more details and a real-world example.

**Review by imagekitio — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (AniVerma17) [reply README.md]: Alright, will change this after discussing this on the call

File: README.md (hunk 5 above)
Comment (AniVerma17) [reply README.md:527]: As per the extensions docs(https://docs.imagekit.io/extensions/overview), the extensions are defined as a list of JSON objects which can have values with any type of primitives, strings or nested JSON objects. One idea could be to use `JsonObject` type, but my thought was to handle it within the SDK and just input the list of maps to reduce verbosity for the consumers. We can discuss this over a call.

File: README.md (hunk 7 above)
Comment (AniVerma17) [reply README.md]: Incorrect text has been fixed, and a brief detail has been added.

**Comment by imagekitio** [comment]
Need to incorporate [format optimisation](https://docs.imagekit.io/features/image-optimization/automatic-image-format-conversion)

**Review by imagekitio — CHANGES_REQUESTED** [review]
Need to incorporate [format optimisation](https://docs.imagekit.io/features/image-optimization/automatic-image-format-conversion)

File: README.md (hunk 1 above)
Comment (AniVerma17) [reply README.md]: Done

File: README.md (hunk 6 above)
Comment (AniVerma17) [reply README.md:566]: Done

File: README.md (hunk 4 above)
Comment (AniVerma17) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (imagekitio) [reply README.md:527]: This is pending. Please share alternative syntax.

Comment (AniVerma17) [reply README.md:527]: The alternative syntax for `extensions` argument, with `JsonArray` as parameter type would be like this:
```
//Kotlin
...,
extensions = JSONArray(arrayOf(
    JSONObject().apply {
        put("name", "google-auto-tagging")
        put("minConfidence", 80)
        put("maxTags", 10)
    }
))
```

**Comment by imagekitio** [comment]
> Need to incorporate [format optimisation](https://docs.imagekit.io/features/image-optimization/automatic-image-format-conversion)

**Comment by AniVerma17** [comment]
> Need to incorporate [format optimisation](https://docs.imagekit.io/features/image-optimization/automatic-image-format-conversion)

For now, we can go with the base approach discussed i.e. determine the optimal format according to the Android version at runtime. We can create a method in the URL constructor, where the developer can pass a boolean flag to explicitly set the automatic format selection by the SDK.
@imagekitio please share your opinion on this.

**Pull request merged**